URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
Commit id: ed4b7a4

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.preprocessing import PolynomialFeatures



## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

url = 'science_data_large.csv'
# Load the dataset
data = pd.read_csv(url)

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X = data.drop(columns=['Size nm^3'])  # Features
y = data['Size nm^3']                  # Labels

# Split the data into training and test sets (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Output the shapes to confirm the split
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (900, 2)
X_test shape: (100, 2)
y_train shape: (900,)
y_test shape: (100,)


## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

# Initialize the model
model = LinearRegression()

# Train the model on the training set
model.fit(X_train, y_train)



LinearRegression()

In [5]:
# Create a sample datapoint (for example)
sample_datapoint = pd.DataFrame({'Temperature °C': [650], 'Mols KCL': [500]})

# Predict the output using the trained model
predicted_output = model.predict(sample_datapoint)
print("Predicted Output:", predicted_output)


Predicted Output: [669951.22235124]


In [6]:
# Calculate the R^2 score on the training set
train_score = model.score(X_train, y_train)
print("Training set R^2 score:", train_score)

# Optionally, calculate the R^2 score on the test set to evaluate generalization
test_score = model.score(X_test, y_test)
print("Test set R^2 score:", test_score)

Training set R^2 score: 0.8610082946223788
Test set R^2 score: 0.8552472077276095


In this model, the R2 scode is 0.86. The score is on a scale of 0 to 1, with 1 being completely accurate and 0 being completely inaccurate. A score of 0.86 shows the model is fairlty accurate, and works well.

In [7]:
# Get coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

print("coefficients", coefficients)
print("intercept",intercept)

coefficients [ 866.14641337 1032.69506649]
intercept -409391.47958340764


$$h(\text{Size}) = -409391.48 + 866.15 \cdot \text{Temperature}^\circ C + 1032.70 \cdot \text{Mols KCl}$$



## Part 4. Use Cross Validation

In [8]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

# Use cross_val_score to evaluate the model
# Let's say we want to use 10-fold cross-validation
scores = cross_val_score(model, X, y, cv=10, scoring='r2')  # Use 'r2' scoring for regression models

# Report findings
print("Cross-validation scores for each fold:", scores)
print("Mean R^2 score:", scores.mean())
print("Standard Deviation of R^2 scores:", scores.std())


Cross-validation scores for each fold: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Mean R^2 score: 0.8552450341984701
Standard Deviation of R^2 scores: 0.03152876296534241


The cross validation score breakes up the data into multiple sets, I chose 10, and then calculates an R2 score for each one of those sets, figuring out the accuracy. Then it shows that the the mean aligns with our overall accuracy score, showing that different subsets also average out to a similar accuracy. The standard deviation is also fairly low, showing that the R2 score remains quite consistant

## Part 5. Using Polynomial Regression

In [9]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

# Create polynomial features of degree 2 for the training set

# Split features and target variable
X = data.drop(columns=['Size nm^3'])
y = data['Size nm^3']

# Split the original data into training and test sets (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

poly = PolynomialFeatures(degree=2, include_bias=False)  # No bias term since LinearRegression adds it
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Initialize and fit the model on the training set
model_poly = LinearRegression()
model_poly.fit(X_train_poly, y_train)

# Evaluate the model using cross-validation on the training set
scores = cross_val_score(model_poly, X_train_poly, y_train, cv=10, scoring='r2')

# Report the metrics
print("Cross-validation scores for each fold:", scores)
print("Mean R^2 score for polynomial regression:", scores.mean())
print("Standard Deviation of R^2 scores for polynomial regression:", scores.std())

# Get coefficients and intercept for the equation
coefficients = model_poly.coef_
intercept = model_poly.intercept_
print("coefficients", coefficients)
print("intercept",intercept)


Cross-validation scores for each fold: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
Mean R^2 score for polynomial regression: 1.0
Standard Deviation of R^2 scores for polynomial regression: 0.0
coefficients [ 1.20000000e+01 -1.27196641e-07  1.26193500e-11  2.00000000e+00
  2.85714287e-02]
intercept 2.0480656530708075e-05



$$y = 2.0480656530708075 \times 10^{-5} + 12.0 \cdot x_1 - 1.27196641 \times 10^{-7} \cdot x_2 + 1.26193500 \times 10^{-11} \cdot x_3 + 2.0 \cdot x_4 + 0.0285714287 \cdot x_5$$

$$x_1 = \text{Temperature °C} $$
$$x_2 = \text{Mols KC } $$


The results here show that the cross validation score indicates that there is a very high R2 score, which makes almost every prediction completely accurately. 